### DISCUSSION TITLE
유니티 GetComponent 최적화를 위한 캐싱은 필요할까요?

### DISCUSSION THREAD

**Discussion by x4cf** [naverz/zepeto-studio-kor#497] [4. General]
과거부터 꾸준히 나오던 주제이죠.
"연산량을 줄이기 위해 맴버변수로 캐싱을 해야한다"는 주장과 "연산에 큰 영향을 주지 않는다, 개발자의 특성에 맞는 가독성이 중요하다"는 주장이죠.

=1번 캐싱=
<img width="492" alt="캐싱" src="https://user-images.githubusercontent.com/90679926/158498216-20067f22-c06d-444d-a154-26773ae7f991.png">

=2번 No캐싱= 
<img width="492" alt="No캐싱" src="https://user-images.githubusercontent.com/90679926/158498221-963a3b85-5ea1-4a64-acca-83b4280556e8.png">

지금은 "성능에 큰 영향을 미치지는 않기때문에 프로그래머가 사용하기 편한 방법으로 사용한다..." 라고하네요.

2015년 자료 기준으로는 연산량이 두 배까지도 차이나는 문제였는데 유니티는 어떻게 해결했을까요?
그리고 유니티에서 해결한 방법이 유니티의 제페토 스크립트에서도 동일하게 작동할까요?

여러분들은 어떤 방법을 사용하시나요?

유저에게 최고의 경험을 선물하기위해 한번의 연산이라도 줄이고 싶나요?
인간은 완벽할 수 없기때문에 개발 코스트를 조금이라도 줄여야한다고 생각하시나요?

**Comment by Joyounwoo** [comment]
저는 가능하다면 대부분 캐싱해서 사용합니다.
하지만 개발자 입장에서 가독성 또한 놓치고 싶지 않은 것도 사실이죠.
많이 호출되지 않는다면 GetComponent를 사용하여 멤버변수의 불필요한 확장을 줄이고,
빈번히 접근하는 경우 멤버변수를 사용하고 있습니다.


(이 주제는 최적화라는 주제와도 맞닿아 있다고도 생각되어 부수적으로 같이 적어 봅니다.)

하지만 요즘처럼 기기의 성능이 매우 향상된 시대에서는 최적화를 위해 소모되는 **개발 자원**에 비해 **효율**이 떨어지는 작업은 필요치 않다고 생각합니다.
물론, 높은 수준의 그래픽 퍼포먼스를 보여주거나 다량의 연산량이 요구되는 프로젝트의 경우 작은 것이라도 최적화가 중요할 것이라 생각합니다. 이 경우에는 가능한 최상의 경험을 유저에게 선사하고자 할 테고, 더불어 개발사에는 고급인력을 포함한 많은 개발 인력이 투입되기에 최적화 작업의 효용성이 높다고 봅니다.

결국 개발하는 프로젝트의 성향과 목적에 따라 달라진다고 생각합니다.
개발 인력이 적고, 성능 이슈를 신경쓸 필요가 없다면 개발자가 관리하기 쉬운 형태로 개발해도 좋다고 생각합니다.
개발엔진을 통한 개발의 장점 또한 이런 곳에 있는 것이니까요.

다만 제페토 월드의 경우, 모바일 환경에 멀티플레이도 가능하며 플레이어 캐릭터의 모델이 별도로 제공되기에 가능하다면 최적화 하는 것이 타당하다고 생각합니다.
다양한 환경의 유저가 가능한 많이, 원활한 접근이 가능한게 좋지 않을까 싶습니다.
